sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1704 @@
+      SpillFramework.withHostSpillBounceBuffer { hmb =>
+        val bbLength = hmb.getLength.toInt
+        withResource(new HostMemoryOutputStream(hmb)) { out =>
+          var sizeRead = IOUtils.copyLarge(in, out, 0, bbLength)
+          while (sizeRead > 0) {
+            // this syncs at every copy, since for now we are
+            // reusing a single host spill bounce buffer
+            dmb.copyFromHostBuffer(
+              /*dstOffset*/ copyOffset,
+              /*src*/ hmb,
+              /*srcOffset*/ 0,
+              /*length*/ sizeRead)
+            out.seek(0) // start over
+            copyOffset += sizeRead
+            sizeRead = IOUtils.copyLarge(in, out, 0, bbLength)
+          }
+        }
+      }
+    }
+  }
+}
+
+trait HandleStore[T <: StoreHandle] extends AutoCloseable with Logging {
+  protected val handles = new ConcurrentHashMap[T, java.lang.Boolean]()
+
+  def numHandles: Int = {
+    handles.size()
+  }
+
+  def track(handle: T): Unit = {
+    doTrack(handle)
+  }
+
+  def remove(handle: T): Unit = {
+    doRemove(handle)
+  }
+
+  protected def doTrack(handle: T): Boolean = {
+    handles.put(handle, true) == null
+  }
+
+  protected def doRemove(handle: T): Boolean = {
+    handles.remove(handle) != null
+  }
+
+  override def close(): Unit = {
+    handles.forEach((handle, _ )=> {
+      handle.close()
+    })
+    handles.clear()
+  }
+}
+
+trait SpillableStore extends HandleStore[SpillableHandle] with Logging {
+  protected def spillNvtxRange: NvtxRange
+
+  def spill(spillNeeded: Long): Long = {
+    if (spillNeeded == 0) {
+      0L
+    } else {
+      withResource(spillNvtxRange) { _ =>
+        var amountSpilled = 0L
+        val spillables = new util.ArrayList[SpillableHandle]()
+        var amountToSpill = 0L
+        val allHandles = handles.keySet().iterator()
+        // two threads could be here trying to spill and creating a list of spillables
+        while (allHandles.hasNext && amountToSpill < spillNeeded) {
+          val handle = allHandles.next()
+          if (handle.spillable) {
+            amountToSpill += handle.approxSizeInBytes
+            spillables.add(handle)
+          }
+        }
+        val it = spillables.iterator()
+        var numSpilled = 0
+        while (it.hasNext) {
+          val handle = it.next()
+          val spilled = handle.spill()
+          if (spilled > 0) {
+            // this thread was successful at spilling handle.
+            amountSpilled += spilled
+            numSpilled += 1
+          } else {
+            // else, either:
+            // - this thread lost the race and the handle was closed
+            // - another thread spilled it
+            // - the handle isn't spillable anymore, due to ref count.
+            it.remove()
+          }
+        }
+        // spillables is the list of handles that have to be closed
+        // we synchronize every thread before we release what was spilled
+        Cuda.deviceSynchronize()
```

Comment:
This is specific to device stores.  We don't need to synchronize the device when spilling from host to disk, for example.